[DPE-2082] User-friendly output when no primary account is defined

deusebio opened the pull request.

juditnovak committed:
CI/CD workflows (#2)

This PR is both introducing CI-CD pipelines for the project, and it is providing a bridge from "snap to lib" (commit
https://github.com/canonical/spark-k8s-toolkit-py/pull/2/commits/b46808ebf25c7bd0c6f287b1d6a778d417c0885d specifically). In case if preferred, we can split those two functional changes in two PRs, it's trivial to do on my side.

NOTE: I couldn't test the Github issue to Jira workflow (the issue I've created doesn't seem to appear in Jira). I'd like to ask for help with that, in case it's to be investigated.

* (origin/main, origin/HEAD, main) [DPE-1858] Migrate
* Increasing line length boundary in pyproject.toml
* Setting initial vesion in pyproject.toml
* microk8s group changes to take impact
* Pinning down local directory definition (no snap environment)
* CI workflow on_pull_request
* Workflow for release 
* Workflow for Jira ticket creation from bug/issue
* Unittests moved out from 'make checks'

juditnovak committed:
Reanaming the spark_client package to spark8t (#4)

Reanaming the spark_client package to spark8t

deusebio committed:
[DPE-1989] Remove spark-client references from spark8t (#7)

deusebio committed:
[DPE-1952] Remove snap references (#5)

deusebio committed:
[DPE-1988] Introduce `backend` arugment in CLI (#6)

juditnovak committed:
Removing confs file (#11)

carlcsaposs-canonical committed:
Update bug_report.md issue template (#12)

Ported from
https://github.com/canonical/mysql-router-k8s-operator/pull/91

deusebio committed:
[DPE-2082] User-friendly output when no primary account is defined

juditnovak commented on spark8t/cli/pyspark.py:
```diff
@@ -38,10 +36,27 @@
 
     if service_account is None:
-        raise ValueError("Service account provided does not exist.")
+        raise NoAccountFound()
```
Why don't we just stop execution and simply return nothing i.e. empty result?

That would be consistent with the behavior of the API.

deusebio committed:
[DPE-2087] Improve logging handling (#15)

juditnovak committed:
[DPE-2095] integration tests to pytest (#16)

Blindly moving from unittest to pytest, as a pre-stage to complete
[DPE-2021](https://warthogs.atlassian.net/browse/DPE-2021).
Related, dependent PR is
https://github.com/canonical/spark-k8s-toolkit-py/pull/17

Stateful, "cleaning up after myself" tests not used anymore.
Instead we have reliable fixtures always ensuring a clean state (in
terms of namespaces).
(Note: So far I see no desperate need to clean up the `default`
namespace, but can be added if needed)

Furthermore `tests/__init__.py` is cleaned up (none of the content was
needed there).

[DPE-2021]:
https://warthogs.atlassian.net/browse/DPE-2021?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

While it may be difficult to review changes, note that test coverage is
equal to[ the last one before this
PR](https://github.com/canonical/spark-k8s-toolkit-py/actions/runs/5253123178):

```
---------- coverage: platform linux, python 3.9.16-final-0 -----------
Name                                      Stmts   Miss  Cover
-------------------------------------------------------------
spark8t/__init__.py                           0      0   100%
spark8t/cli/__init__.py                       3      3     0%
spark8t/cli/params.py                        29     29     0%
spark8t/cli/pyspark.py                       15     15     0%
spark8t/cli/service_account_registry.py      76     76     0%
spark8t/cli/spark_shell.py                   15     15     0%
spark8t/cli/spark_submit.py                  15     15     0%
spark8t/domain.py                           161     27    83%
spark8t/exceptions.py                        10      1    90%
spark8t/literals.py                           3      0   100%
spark8t/services.py                         527    202    62%
spark8t/utils.py                             93     27    71%
-------------------------------------------------------------
TOTAL                                       947    410    57%

================== 12 passed, 2 warnings in 138.45s (0:02:18) ==================
  integration: OK (144.86=setup[1.35]+cmd[3.22,0.59,139.71] seconds)
  congratulations :) (144.99 seconds)
```

[New
execution](https://github.com/canonical/spark-k8s-toolkit-py/actions/runs/5261122237)
```
---------- coverage: platform linux, python 3.9.17-final-0 -----------
Name                                      Stmts   Miss  Cover
-------------------------------------------------------------
spark8t/__init__.py                           0      0   100%
spark8t/cli/__init__.py                       3      3     0%
spark8t/cli/params.py                        29     29     0%
spark8t/cli/pyspark.py                       15     15     0%
spark8t/cli/service_account_registry.py      76     76     0%
spark8t/cli/spark_shell.py                   15     15     0%
spark8t/cli/spark_submit.py                  15     15     0%
spark8t/domain.py                           161     27    83%
spark8t/exceptions.py                        10      1    90%
spark8t/literals.py                           3      0   100%
spark8t/services.py                         527    202    62%
spark8t/utils.py                             93     27    71%
-------------------------------------------------------------
TOTAL                                       947    410    57%

================== 12 passed, 2 warnings in 133.44s (0:02:13) ==================
  integration: OK (142.51=setup[1.63]+cmd[3.67,1.18,136.03] seconds)
  congratulations :) (142.66 seconds)
```

juditnovak committed:
[DPE-2021] integration tests lightkube registry (#17)

[DPE-2021] integration tests lightkube registry

juditnovak committed:
[DPE-2131][BUGFIX] lightkube rolebindnig broken (#19)

On namespace creation the follwing error ocurred:

```
FAILED tests/integration/test_registry.py::TestRegistry::test_registry_change_primary_account_1_spark_namespace - lightkube.core.exceptions.ApiError: RoleBinding.rbac.authorization.k8s.io "spark-user-1-role-binding" is invalid: subjects[0].name: Required value
```
* LightKube service fixed
* Tests enabled
* Unittest workflow changes

deusebio committed:
handle NotResourceFound exceptions nicely

deusebio committed:
fix linting

deusebio committed:
missing resource handling for lightkube

deusebio committed:
fix linting

juditnovak committed:
[DPE-2102] unittest to pytest (#18)

Moving unittests to pytest

[Pipeline before unittest
modifications](https://github.com/canonical/spark-k8s-toolkit-py/actions/runs/5253123178)

I'm slightly unsure what causes the 3-lines difference on the 1500+
lines test module `test_services.py`... But if it's preferred, I can try
to find the difference. (If it's sufficiently small not to worry, then
I'm also good to leave it.)

```
---------- coverage: platform linux, python 3.9.16-final-0 -----------
Name                                      Stmts   Miss  Cover
-------------------------------------------------------------
spark8t/__init__.py                           0      0   100%
spark8t/cli/__init__.py                       3      0   100%
spark8t/cli/params.py                        29      1    97%
spark8t/cli/pyspark.py                       15     15     0%
spark8t/cli/service_account_registry.py      76     45    41%
spark8t/cli/spark_shell.py                   15     15     0%
spark8t/cli/spark_submit.py                  15     15     0%
spark8t/domain.py                           161      7    96%
spark8t/exceptions.py                        10      2    80%
spark8t/literals.py                           3      0   100%
spark8t/services.py                         527    143    73%
spark8t/utils.py                             93     16    83%
-------------------------------------------------------------
TOTAL                                       947    259    73%

======================= 53 passed, 2 warnings in 23.55s ========================
  unit: OK (30.69=setup[1.15]+cmd[2.94,0.56,26.03] seconds)
  congratulations :) (31.23 seconds)
```


[Current
pipeline](https://github.com/canonical/spark-k8s-toolkit-py/actions/runs/5269090394)
```
---------- coverage: platform linux, python 3.9.16-final-0 -----------
Name                                      Stmts   Miss  Cover
-------------------------------------------------------------
spark8t/__init__.py                           0      0   100%
spark8t/cli/__init__.py                       3      0   100%
spark8t/cli/params.py                        29      1    97%
spark8t/cli/pyspark.py                       15     15     0%
spark8t/cli/service_account_registry.py      76     45    41%
spark8t/cli/spark_shell.py                   15     15     0%
spark8t/cli/spark_submit.py                  15     15     0%
spark8t/domain.py                           161      7    96%
spark8t/exceptions.py                        10      2    80%
spark8t/literals.py                           3      0   100%
spark8t/services.py                         527    146    72%
spark8t/utils.py                             93     16    83%
-------------------------------------------------------------
TOTAL                                       947    262    72%

======================= 53 passed, 2 warnings in 16.23s ========================
  unit: OK (23.84=setup[1.48]+cmd[3.48,0.75,18.12] seconds)
  congratulations :) (24.29 seconds)
  ```

* pytest-mock added
* Adding tests/unittest/conftest.py
* test_services.py to pytest
* test_domain.py to unittest
* test_argument_parsing to pytest

juditnovak committed:
[DPE-1990] README (#14)


---------

Co-authored-by: Enrico Deusebio <enrico.deusebio@canonical.com>

deusebio committed:
[DPE-2085] Error on unrecognizable user input (#22)

deusebio committed:
address code-review

deusebio committed:
poetry update

deusebio committed:
fix linting
